File: keejob store-angular/src/app/models/formation-keejob.ts
import { Plateforme } from "./platforme";



export enum FormationCategory {
  Formations_langues = 'Formations_langues',
  Formations_office = 'Formations_office',
  Formations_Design = 'Formations_Design',
  Formations_Digital = 'Formations_Digital',

}



export interface Avantage {
  titre: string;
}

export interface FormationKeejob {
  id?: number;

  titre: string;
  badge: string;
  image: string;                  // URL Cloudinary
  lienFormation: string;

  descriptionCourte: string;      // max 500 caractères
  aPropos: string;                // max 2000 caractères

  // note: number;
  // nombreAvis: number;
  nombreApprenants: string;

  niveau: string;
  duree: string;
  langue: string;
  sousTitres: string;
  acces: string;
  derniereMiseAJour: string;

  prix: number;
  prixOriginal: number;
  reduction: number;

  accesVie: boolean;
  certificatInclus: boolean;
  garantieRemboursement: string;

  categoryFormationKeejob: FormationCategory;

  avantages: Avantage[];
  competencesAcquises: string[];
  publicCible: string[];

  plateforme?: Plateforme | { id: number };

  // avis est marqué @JsonIgnore côté back → généralement non exposé
  // avis?: Avis[];
}

File: keejobStore/src/main/java/com/example/keejobstore/entity/FormationKeejob.java
package com.example.keejobstore.entity;

import com.example.keejobstore.converter.AvantagesFormationConverter;
import com.example.keejobstore.converter.CompetencesFormationConverter;
import com.example.keejobstore.converter.PublicCibleFormationConverter;
import com.fasterxml.jackson.annotation.JsonIgnore;
import jakarta.persistence.*;
import lombok.*;
import lombok.experimental.FieldDefaults;

import java.util.ArrayList;
import java.util.List;

@Setter
@FieldDefaults(level = AccessLevel.PRIVATE)
@Getter
@AllArgsConstructor
@NoArgsConstructor
@Entity
public class FormationKeejob {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    private String titre;
    private String badge;
    private String image;                  // URL Cloudinary
    //private String lienBandeAnnonce;
    private String lienFormation;

    @Column(length = 500)
    private String descriptionCourte;

    @Column(length = 2000)
    private String aPropos;


    private String nombreApprenants;

    private String niveau;
    private String duree;
    private String langue;
    private String sousTitres;
    private String acces;
    private String derniereMiseAJour;

    private Double prix;
    private Double prixOriginal;
    private Integer reduction;

    private boolean accesVie;
    private boolean certificatInclus;
    private String garantieRemboursement;

    @Enumerated(EnumType.STRING)
    private CategoryFormationKeejob categoryFormationKeejob;   // ← réintégré, utilisé par la navigation du site

    @Convert(converter = AvantagesFormationConverter.class)
    @Column(columnDefinition = "JSON")
    private List<Avantage> avantages = new ArrayList<>();

    @Convert(converter = CompetencesFormationConverter.class)
    @Column(columnDefinition = "JSON")
    private List<String> competencesAcquises = new ArrayList<>();

    @Convert(converter = PublicCibleFormationConverter.class)
    @Column(columnDefinition = "JSON")
    private List<String> publicCible = new ArrayList<>();

    @ManyToOne
    @JoinColumn(name = "plateforme_id")
    private Plateforme plateforme;

    @OneToMany(mappedBy = "formation", cascade = CascadeType.ALL, orphanRemoval = true)
    @JsonIgnore
    private List<Avis> avis = new ArrayList<>();

    @Setter
    @FieldDefaults(level = AccessLevel.PRIVATE)
    @Getter
    @AllArgsConstructor
    @NoArgsConstructor
    public static class Avantage {
        private String titre;
    }
}